R/model_ivpin.R

```diff
@@ -0,0 +1,905 @@
+
+    duration <- as.numeric(duration)
+
+    if (days > 0) {
+      dseconds <- (days - 1) * 3600 * 24 + (24 - tradinghours) * 3600
+      return(duration - dseconds)
+    }
+    return(duration)
+  }
+
+  if (tradinghours > 0 & tradinghours < 24) {
+    bucketdata$duration <- apply(bucketdata, 1, function(x)
+      corduration(x[6], x[5], x[10]))
+  }
+  print(nrow(data))
+  # Normalize the duration to be centered around 1
+  mean_duration <- mean(bucketdata$duration)
+  bucketdata$duration <- bucketdata$duration / mean_duration
+  bucketdata$duration <- ifelse(bucketdata$duration == 0, 0.1, bucketdata$duration)
+                                
+  compute_log_likelihood <- function(params, t, Vb, Vs) {
+    alpha <- sigmoid(params[1])
+    delta <- sigmoid(params[2])
+    mu <- params[3] ^ 2
+    eps <- params[4] ^ 2
+    
+    e_1 <- log(alpha * delta) + Vb * log(eps) + Vs * log(eps + mu) - (2 * eps + mu) * t
+    e_2 <- log(alpha * (1 - delta)) + Vb * log(eps + mu) + Vs * log(eps) - (2 * eps + mu) * t
+    e_3 <- log(1 - alpha) + Vb * log(eps) + Vs * log(eps) - 2 * eps * t
+    e_max <- pmax(e_1, e_2, e_3)
+    
+    log_likelihood <- -sum(log(exp(e_1 - e_max) + exp(e_2 - e_max) + exp(e_3 - e_max)) + e_max)
+    return(log_likelihood)
+  }
+
+  # Compute iVPIN with a rolling window
+
+  start_index <- samplength + 1
+
+  "
+  Easley et al. (2012b) derive the VPIN estimator based on the argument of two moment conditions, 
+  E[|VτB - VτS|] ≈ αμ and E[|VτB + VτS|] = 2ε + αμ, from the Poisson processes. 
+  According to Ke et al. (2017), the two moment conditions should instead be expressed as
+  E[|VτB-VτS||tτ;θ]≈ αμtτ and E[|VτB + VτS|tτ; θ ] = (2ε + αμ)tτ,
+  "
+  total_arrival_rate <- (bucketdata$agg.bvol + bucketdata$agg.svol) / bucketdata$duration
+  informed_arrival <- abs(bucketdata$agg.bvol - bucketdata$agg.svol) / bucketdata$duration
+
+  t <- bucketdata$duration
+  Vb <- bucketdata$agg.bvol
+  Vs <- bucketdata$agg.svol
+
+  ivpin <- numeric(nrow(bucketdata))  # Ensure ivpin is initialized
+
+  for (i in start_index:nrow(bucketdata)) {
+    j <- i - samplength
+    parms <- rep(-Inf, 4)  # alpha, delta, mu, eps
+    
+    log_lik <- Inf
+    flag <- -Inf
+    
+    best_log_likelihood <- Inf
+    exit_flag <- FALSE
+    
+    if (i == start_index || i %% 10 == 0) {
+      for (alpha_init in seq(0.1, 0.9, by = 0.2)) {
+        for (delta_init in seq(0.1, 0.9, by = 0.2)) {
+          mu_init <- mean(informed_arrival[j:i]) / alpha_init
+          eps_init <- mean(abs(total_arrival_rate[j:i] - informed_arrival[j:i])) / 2
+          
+          initial_guess <- c(logit(max(min(alpha_init, 0.999), 0.001)), 
+                             logit(max(min(delta_init, 0.999), 0.001)), 
+                             sqrt(mu_init), 
+                             sqrt(eps_init))
+          tryCatch({
+            result <- optim(initial_guess, compute_log_likelihood, t = t[j:i], Vb = Vb[j:i], Vs = Vs[j:i], method = "BFGS")
```

Comment:
In the package, use the package `nloptr` for nonlinear optimization, please use the function `nloptr::lbfgs()` instead.